### ISSUE TITLE
Parent/child: Change structure of has_child, has_parent queries and filters.

### ISSUE THREAD

**Issue by martijnvg** [elastic/elasticsearch#8154]
Currently the structure of the `has_child` query is:
```json
{
   "has_child" : {
      "type" : "a-child-type",
      "query ": { <query> } 
   }
}
```

If the structure were to be changed to:

```json
{
   "has_child" : {
      "a-child-type" : {
         "query" : { <query> } 
      }  
   }
}
```

Then the structure is streaming parsing friendly and the child type is always known when the `query` is parsed. 

At the moment if the `query` field is encountered before the `type` field by the parser then the query json object is read into memory, but not parsed. It is parsed after the `type` field has been read. This work around can then also be removed.

Also validation should be added whether the used type is valid to use. This becomes useful in cases where a query contains multiple hierarchal `has_child` or `has_parent` queries. For example if the following types are defined grand_parent -> parent -> child. Then the only valid type when wrapping a has_child query in a has_child query that has type `parent` is `child`. This validation is already possible, but it is a nice opportunity to add this as well with this change.

**Comment by s1monw** [comment]
I think we should do this and also make sure we fix #8155 for consistency. @martijnvg I guess we have to provide bwc for this unfortunately

**Comment by martijnvg** [comment]
Yes, bwc for this is unfortunately, but it seems the only way forward.

The bwc logic is going to be tricky. (there may be a type with the name
`query`) We can only make the decision if the old or new syntax is used the
two fields (`query` and `type`) either do or do not exist under the
`has_child` or `has_parent`. I think this bwc logic prevent streaming
parsing for the time being...

On 20 February 2015 at 10:38, Simon Willnauer <notifications@github.com>
wrote:

> I think we should do this and also make sure we fix #8155
> <https://github.com/elasticsearch/elasticsearch/issues/8155> for
> consistency. @martijnvg <https://github.com/martijnvg> I guess we have to
> provide bwc for this unfortunately
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elasticsearch/elasticsearch/issues/8154#issuecomment-75211771>
> .
>



-- 
Met vriendelijke groet,

Martijn van Groningen

**Comment by martijnvg** [comment]
Can be closed as it is no longer needed for the same reason as in: https://github.com/elastic/elasticsearch/issues/8155#issuecomment-151028678
